### PR TITLE
#45: Add reusable workflows

### LINKED ISSUE #45: Reusing workflows
I just found out that [reusable workflows](https://docs.github.com/en/actions/sharing-automations/reusing-workflows) exist, which I think would be good for us. In all 3 repos we have the identical `citation_check.yml` workflow and an almost identical `test_suite.yml`. So it's a bit of a hassle that we have to edit all 3 when making changes. We could host these common workflows in `mesh-adaptation-docs` and make smaller workflows in the 3 repos that would call them. What do you think @jwallwork23?

### PR DESCRIPTION
**Pull request by ddundo**
Closes #45.

After this is merged I will change `citation_check.yml` in Animate/Goalie/Movement to be:
```
name: 'Citation Check'

on:
  push:
    branches:
      - main
    paths:
      - '**/*.cff'

  pull_request:
    paths:
      - '**/*.cff'

  workflow_dispatch:

jobs:
  citation_check:
    uses: mesh-adaptation/mesh-adaptation-docs/.github/workflows/citation_check.yml@main
```

and `test_suite.yml` in Animate will be:
```
name: 'Run Animate Test Suite'

on:
  push:
    branches:
      - main
  pull_request:
  schedule:
    - cron: '0 1 * * 0'

jobs:
  test_suite:
    uses: mesh-adaptation/mesh-adaptation-docs/.github/workflows/test_suite.yml@main
    with:
      install-command: 'python -m pip uninstall -y animate && python -m pip install -e .'
      test-command: |
        export GITHUB_ACTIONS_TEST_RUN=1
        python $(which firedrake-clean)
        export ANIMATE_CHECKPOINT_DIR=$(pwd)/.checkpoints
        python -m coverage erase
        python -m coverage run --source=animate -m pytest -v --durations=20 test
        python -m coverage report
      changed-files-patterns: |
        **/*.py
        **/*.cxx
```

and similarly for Goalie and Movement.

### CODE REVIEW

File: .github/workflows/reusable_citation_check.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: 'Check out the repo'
+        uses: actions/checkout@v4
+
+      - name: 'Setup Python'
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.8'
```
Comment (jwallwork23) [inline .github/workflows/reusable_citation_check.yml:18]: While we're at it maybe we upgrade to a more recent Python?

File: .github/workflows/reusable_test_suite.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        if: ${{ always() }}
+        run: |
+          cd ..
+          rm -rf build
+
+      - name: 'Setup Python'
+        if: ${{ (github.event_name == 'push' && github.ref == 'refs/heads/main') || (steps.changed-files.outcome == 'success' && steps.changed-files.outputs.any_changed == 'true') || github.event_name == 'schedule' }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.8'
```
Comment (jwallwork23) [inline .github/workflows/reusable_test_suite.yml:51]: Here too.

**Review by jwallwork23 — APPROVED** [review]
This is cool, thanks @ddundo. Just one minor suggestion.

File: .github/workflows/reusable_citation_check.yml (hunk 1 above)
Comment (ddundo) [reply .github/workflows/reusable_citation_check.yml:18]: Sure! 3.12 or something older? :)

Comment (ddundo) [reply .github/workflows/reusable_citation_check.yml:18]: @jwallwork23 ping for notification :)

Comment (jwallwork23) [reply .github/workflows/reusable_citation_check.yml:18]: > Sure! 3.12 or something older? :)

Yeah give that a go.

**Comment by ddundo** [comment]
@jwallwork23 maybe I could also remove
```
check_citation:
	@echo "Checking CITATION.cff formatting..."
	@cffconvert --validate
	@echo "PASS"
```

from the Makefiles and add it centrally to `reusable_citation_check.yml`?

File: .github/workflows/reusable_citation_check.yml (hunk 1 above)
Comment (jwallwork23) [reply .github/workflows/reusable_citation_check.yml:18]: > @jwallwork23 ping for notification :)

No need to do this any more btw - I turned on all notifications for the mesh-adaptation repos.

**Comment by jwallwork23** [comment]
> @jwallwork23 maybe I could also remove
> 
> ```
> check_citation:
> 	@echo "Checking CITATION.cff formatting..."
> 	@cffconvert --validate
> 	@echo "PASS"
> ```
> 
> from the Makefiles and add it centrally to `reusable_citation_check.yml`?

I'd prefer to keep it in so that users and developers can run locally.

**Comment by jwallwork23** [comment]
> @jwallwork23 maybe I could also remove
> 
> ```
> check_citation:
> 	@echo "Checking CITATION.cff formatting..."
> 	@cffconvert --validate
> 	@echo "PASS"
> ```
> 
> from the Makefiles and add it centrally to `reusable_citation_check.yml`?

Actually, you're right - it's probably not necessary for this to be duplicated across all the repos. The citation isn't going to change that often either. Let's remove it from the Makefile.

File: .github/workflows/reusable_citation_check.yml (hunk 1 above)
Comment (ddundo) [reply .github/workflows/reusable_citation_check.yml]: Done in d4c059d1dfb3abdb24e2686b353598950bfe64d5

File: .github/workflows/reusable_test_suite.yml (hunk 2 above)
Comment (ddundo) [reply .github/workflows/reusable_test_suite.yml]: Done in d4c059d1dfb3abdb24e2686b353598950bfe64d5

**Comment by ddundo** [comment]
Great, thanks! I made these changes and tested it on Animate https://github.com/mesh-adaptation/animate/pull/148 so I'll merge :)

**Pull request merged**